core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -141,7 +141,36 @@ bool cuda_launch_blocking() {
 
 }  // namespace
 
-void cuda_device_synchronize() { CUDA_SAFE_CALL(cudaDeviceSynchronize()); }
+void cuda_device_synchronize(const std::string &name) {
+  Kokkos::Tools::Experimental::Impl::profile_fence_event<Kokkos::Cuda>(
+      name,
+      Kokkos::Tools::Experimental::SpecialSynchronizationCases::
+          GlobalDeviceSynchronization,
+      []() {  // TODO: correct device ID
+        CUDA_SAFE_CALL(cudaDeviceSynchronize());
+      });
+}
+
+void cuda_stream_synchronize(const cudaStream_t stream, const CudaInternal *ptr,
+                             const std::string &name) {
+  Kokkos::Tools::Experimental::Impl::profile_fence_event<Kokkos::Cuda>(
+      name,
+      Kokkos::Tools::Experimental::Impl::DirectFenceIDHandle{
+          ptr->impl_get_instance_id()},
```

Comment:
I am slightly unhappy that this kind of work (looking up ID's etc.) happens even if no profiling tool is loaded, but the benchmarks I ran don't seem to indicate too much impact. Back to back fences are now about 20% slower, but that is from a very low base of about half a microsecond on CUDA. having a empty kernel in between we loose roughly the same absolute time of 0.1us but on a total time of 5us. So we are talking about 2% slow down for a pure cached latency benchmark. What I don't know is whether the time goes up significantly if the underlying map is not in cache anymore. On the other hand if its not in cache anymore that means you moved enough data to evict it, so the time between fences should be substantially larger, reducing the overhead impact.

Reply:
So, I have a way around this, if you'd like to see it explored. Essentially, rather than pass the value, I could pass a function that returns the value, and only invoke the function if a tool is loaded. My suspicion is that this is getting a little baroque for these purposes, but if you want me to check that out I can

Reply:
Here's where I've landed on this: I think we should unify all of the "only do this if a tool is loaded" stuff into a new interface that I have in mind. If we don't think this overhead is catastophic, I'd recommend we merge as is, and I'll take another stab at basically a "lazily acquire callback values" system.